File: 3217/solution.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function modifiedList(nums: number[], head: ListNode | null): ListNode | null {
  const badValues = new Set(nums);
  const dummy = new ListNode();
  let curr = dummy;
  while (head) {
    if (!badValues.has(head.val)) {
      curr.next = head;
      curr = curr.next;
    }
    head = head.next;
    curr.next = null;
  }
  return dummy.next;
}

File: 3217/solution.java
import java.util.HashSet;
import java.util.Set;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        Set<Integer> badValues = new HashSet<>();
        for (int num : nums) {
            badValues.add(num);
        }
        ListNode dummy = new ListNode();
        ListNode curr = dummy;
        while (head != null) {
            if (!badValues.contains(head.val)) {
                curr.next = head;
                curr = curr.next;
            }
            head = head.next;
            curr.next = null;
        }
        return dummy.next;
    }
}

// time O(n)
// space O(1)